foxpuppet/windows/browser/tab_bar.py

```diff
@@ -0,0 +1,98 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/
+"""Creates tab_bar object to interact with the Firefox tabs and tabbar."""
+
+from selenium.webdriver.common.by import By
+
+from foxpuppet.region import Region
+
+
+class Tab_bar(Region):
+    """Representation of the tab bar which contains the tabs.
+
+    Args:
+        window (:py:class:`BaseWindow`): Window object this region appears in.
+        root
+            (:py:class:`~selenium.webdriver.remote.webelement.WebElement`):
+            WebDriver element object that serves as the root for the
+            region.
+
+    """
+
+    _new_tab_button_locator = (By.ID, 'new-tab-button')
+    _tabs_locator = (By.TAG_NAME, 'tab')
+
+    @property
+    def tabs(self):
+        """Return a list of tabs.
+
+        Returns: :py:class:`~foxpuppet.window.browser.tab_bar.Tab`
+
+        """
+        with self.selenium.context(self.selenium.CONTEXT_CHROME):
+            tabs = [self.Tab(self, el) for el in
+                    self.selenium.find_elements(*self._tabs_locator)]
+        # Assign handles
+        for tab, handle in zip(tabs, self.selenium.window_handles):
+            tab.handle = handle
+        return tabs
+
+    def open_new_tab(self):
+        """Open a new tab in the current window.
+
+        Returns: list of :py:class:`Tab`.
+
+        """
+        current_tabs = len(self.selenium.window_handles)
+        with self.selenium.context(self.selenium.CONTEXT_CHROME):
+            self.selenium.find_element(*self._new_tab_button_locator).click()
+        self.wait.until(
+            lambda _: len(self.selenium.window_handles) != current_tabs)
+        return self.tabs
+
+    class Tab(Region):
+        """Representaion of the Tab.
+
+        Args:
+            window (:py:class:`BaseWindow`): Window object this region appears
+                    in.
+            root
+                (:py:class:`~selenium.webdriver.remote.webelement.WebElement`):
+                WebDriver element object that serves as the root for the
+                region.
+
+        """
+
+        def close(self):
+            """Close the selected tab."""
+            current_tabs = len(self.selenium.window_handles)
+            with self.selenium.context(self.selenium.CONTEXT_CHROME):
+                button = self.root.find_anonymous_element_by_attribute(
+                    'anonid', 'close-button')
+                button.click()
+            self.wait.until(
+                lambda _: len(self.selenium.window_handles) != current_tabs)
+
+        def select(self):
+            """Select the tab.
+
+            Returns: :py:class:`Tab`.
+
+            """
+            self.selenium.switch_to.window(self.handle)
```

Comment:
This won't show the tab if is not the current tab. In this case, we would want to click on the tab. It's also possible that the tab is scrolled out of view on the tab bar, so we'd need to bring it into view before clicking it.

Reply:
Do we really need to bring it into view if selenium knows where it is? I mean technically it is there, its not hidden or anything.

Reply:
I don't know if you'll be able to click on an element that's not in view. We're simulating a user, and a user can't switch to a tab without being able to click it.

Reply:
Note that Marionette currently doesn't know about tabs which haven't been restored via sessionstore. Those don't have a browser, and as such will not be returned.